### PR TITLE
Fix for 45163

### PR DESCRIPTION
**Pull request by Drizshko**

### CODE REVIEW

**Comment by myshevts** [comment]
@Drizshko can you pls elaborate why this change should help to fix sporadic failure of the test

**Comment by Drizshko** [comment]
@myshevts i did run test for hours on my Mac and Ubuntu --gtest_repeat=-1 with no failures. This is not a fix, this is a way to try another implementation, i'd like to re-run  tests on this branch to make sure  if this helps or not. 

My only idea is a difference in std library version or compiler glitch. Maybe i'm wrong.

**Comment by ilya-lavrenov** [comment]
@myshevts do you think this is a valid fix?

**Comment by myshevts** [comment]
> @myshevts do you think this is a valid fix?

I don't think so.
I guess the issue is as in the Cancelation impl:
 let's inspect the test code (comments are mine)

    req.StartAsync();
    req.Cancel(); // **if Cancel happens AFTER the request is already completed, it's cancelation status is NOT reset**
    req.Wait(InferenceEngine::IInferRequest::WaitMode::RESULT_READY);

    req.StartAsync(); // **hence the StartAsync will throw the INFER_CANCELLED (-13)** 


This is whta we see in the tests log:
  20:48:47  [2020-12-16 17:48:47,764] cpuFuncTests INFO: /home/jenkins/agent/workspace/private-ci/ie/build-linux-ubuntu20/b/repos/openvino/inference-engine/tests/functional/plugin/shared/include/behavior/infer_request_cancellation.hpp:89: Failure
  20:48:47  [2020-12-16 17:48:47,764] cpuFuncTests INFO:       Expected: static_cast<int>(InferenceEngine::StatusCode::OK)
  20:48:47  [2020-12-16 17:48:47,764] cpuFuncTests INFO:       Which is: 0
  20:48:47  [2020-12-16 17:48:47,764] cpuFuncTests INFO: To be equal to: waitStatus
  20:48:47  [2020-12-16 17:48:47,764] cpuFuncTests INFO:       **Which is: -13**
 
The flaw is in the following method:

    void Cancel() {
        cancelation_requested = true;
    }

It should check for the request status, and if it is NOT being executed, should either NOT set the cancelation_requested (or maybe throw an execption), like this:

    void Cancel() {
        if (isRequestBusy())
            cancelation_requested = true;
    }

**Comment by Drizshko** [comment]
@myshevts The problem is - request do not start immediately after calling to StartAsync, so if i'll add this check it will always fail. 
From implemented logic - on cancel code sets cancellation variable, on infer it checks if it was set and if so it reset this variable to false first and only after that throws exception. So after Wait it should be false.

So my only idea was to check what  if i'll replace atomic with regular variable. Maybe for some reason it's not set to false.

**Comment by myshevts** [comment]
> From implemented logic - on cancel code sets cancellation variable, on infer it checks if it was set and if so it reset this variable to false first and only after that throws exception. So after Wait it should be false.

The problem is exacty that the variable is tested/reset ONLY inside inference.
So if the inference is already over, it is not reset (and will cancel the next inference) - this is probably ok (but such behaviour of the Cancel should be carefully documented), but this is why the test fails, imho.

**Comment by apankratovantonp** [comment]
Definitely the value of flag should be **atomic** if it is intended to be set from different threads!

@myshevts  the implementation below dose not guarantees that `cancelation_requested` would be set to true only when infer request is busy:
```c++
void Cancel() {
    if (isRequestBusy())
        cancelation_requested = true;
}
```

Here is my vision how to implement behavior specified by the tests. `cancelation_requested` variable should have 3 states:
```c++
enum InferState {Busy, NotBusy, Cancel};
std::atomic<InferState> cancelation_requested;
```
`MKLDNNGraph::Infer(int batch)` should contains cancelation flag guard:
```c++

void MKLDNNGraph::Infer(int batch) {
struct CancelationGuard {
    CancelationGuard(std::atomic<InferState>& cancelation_requested)  :
     _cancelation_requested(cancelation_requested) {
        _cancelation_requested = InferState::Busy;
    }
    ~CancelationGuard() {
        _cancelation_requested = InferState::NotBusy;
    }
    std::atomic<InferState>& _cancelation_requested;
} cancelationGuard(cancelation_requested);
```

Cancelation flag check code in `MKLDNNGraph::Infer()`:
```c++
InferState expected = InferState::Busy;
If (!cancelation_requested.compare_exchange_strong(expected, InferState::Busy) {
    THROW_IE_EXCEPTION << InferenceEngine::details::as_status << InferenceEngine::INFER_CANCELLED;
}
```

`MKLDNNGraph::Cancel()` Inplementation:
```c++
void MKLDNNGraph::Cancel() {
    InferState expected = InferState::Busy;
    cancelation_requested.compare_exchange_strong(expected, InferState::Cancel);
}
```

**Comment by Drizshko** [comment]
@apankratovantonp Thank you for suggested code, looks good, especially CancelationGuard. One thing here, as i said above - after calling StartAsync it will take some time before MKLDNNGraph::Infer will take a place. So if i'll call StartAsync and Cancel right after - Cancel will fail here: cancelation_requested.compare_exchange_strong(expected, InferState::Cancel);  as cancelation_requested will not be  assigned to Busy yet.

**Comment by myshevts** [comment]
@apankratovantonp the only item that I don't really like in your code is touching the atomic variable on every inference step with compare_exchange.

To me moving the Cancel impl to the class holding async (and making sure that it is updated under same mutex as the request's general busy status is changed) would allow to make the cancelation_requested non-atomic.

**Comment by apankratovantonp** [comment]
> @apankratovantonp Thank you for suggested code, looks good, especially CancelationGuard. One thing here, as i said above - after calling StartAsync it will take some time before MKLDNNGraph::Infer will take a place. So if i'll call StartAsync and Cancel right after - Cancel will fail here: cancelation_requested.compare_exchange_strong(expected, InferState::Cancel); as cancelation_requested will not be assigned to Busy yet.

Yes I see and I agree. So as @myshevts says  cancelation logic should be integrated into Asynchronous infer request. So it would be called in the thread that calls StartAsync. 

I think we should change to three state the busy flag from openvino/inference-engine/src/plugin_api/cpp_interfaces/impl/ie_infer_async_request_thread_safe_internal.hpp:

```c++
class AsyncInferRequestThreadSafeInternal : public IAsyncInferRequestInternal {
    std::atomic_bool _isRequestBusy = {false};
```
`mkldnn_async_infer` request should pass the reference to this flag to `mkldnn_infer_request` in `mkldnn_async_infer` constructor using some new API like `mkldnn_infer_request::SetInferenceStateReference(std::atomic<InferState>& inferState)`

@myshevts We can use any name. And we can add `CancelationGuard::reset()` method and call it in proper place. 

As for performance - It is the payoff for correct asynchronous code. If we integrate it into `AsyncInferRequestThreadSafeInternal` there would be just new `CAS` for each node in MKLDNNGraph. The performance tests should show what is more efficient  `CAS` or mutex.

**Comment by Drizshko** [comment]
I agree about atomic. I did remove atomic to see if tests could be fixed, i was thinking maybe atomic implementation in std library on target platform has some issues or compiler. I did run tests on 3 platforms for thousands of iterations with no issues, so i was a bit surprised if fail on test platform. 

I will try to implement suggested approach.

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/skip_tests_config.cpp [hunk 1]
```diff
@@ -60,7 +60,7 @@ std::vector<std::string> disabledTestPatterns() {
         // TODO: Issue 43417 sporadic issue, looks like an issue in test, reproducible only on Windows platform
         R"(.*decomposition1_batch=5_hidden_size=10_input_size=30_.*tanh.relu.*_clip=0_linear_before_reset=1.*_targetDevice=CPU_.*)",
         // TODO: Sporadic Issue: 45163
-        R"(.*Behavior.*CancellationTests.*canResetAfterCancelAsyncRequest.*)",
+        // R"(.*Behavior.*CancellationTests.*canResetAfterCancelAsyncRequest.*)",
```
Comment (ilya-lavrenov) [inline inference-engine/tests/functional/plugin/cpu/shared_tests_instances/skip_tests_config.cpp:63]: we can just remove these lines.

Do we need changes in MKLDNN submodule?

Comment (Drizshko) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/skip_tests_config.cpp:63]: Looks like committed mkl-dnn accidentally after merging with master

**Pull request closed without merging**